TCC.Tests/Tasks/TaskHelperTest.cs

```diff
@@ -0,0 +1,134 @@
+                await Task.Delay(i, ct);
+                Interlocked.Decrement(ref count);
+                Assert.True(count <= degree);
+            }, degree, Fail.Smart, cts.Token);
+
+            Assert.Equal(degree, max);
+        }
+
+
+        [Theory]
+        [InlineData(Fail.Default)]
+        [InlineData(Fail.Fast)]
+        [InlineData(Fail.Smart)]
+        public async Task CancellationAsync(Fail failMode)
+        {
+            var tasks = Enumerable.Range(0, 10);
+            var cts = new CancellationTokenSource();
+
+            int index = 0;
+
+            var pTask = tasks.ParallelizeAsync(async (i, ct) =>
+            {
+                int ix = Interlocked.Increment(ref index);
+                if (ix == 4)
+                {
+                    cts.Cancel();
+                }
+                else
+                {
+                    await Task.Delay(1000, ct);
+                    if (ct.IsCancellationRequested)
+                    {
+                        return;
+                    }
+                    Assert.False(true);
+                }
+            }, 8, failMode, cts.Token);
+
+            if (failMode == Fail.Default)
+            {
+                await Assert.ThrowsAsync<TaskCanceledException>(async () =>
+                {
+                    await pTask;
+                });
+            }
+            else
+            {
+                var result = await pTask; 
+                Assert.True(result.IsCancelled);
+            }
+        }
+
+        [Theory]
+        [InlineData(Fail.Default)]
+        [InlineData(Fail.Fast)]
+        [InlineData(Fail.Smart)]
+        public async Task ExceptionAsync(Fail failMode)
+        {
+            var tasks = Enumerable.Range(0, 10);
+            int index = 0;
+
+            var pTask = tasks.ParallelizeAsync(async (i, ct) =>
+            {
+                int ix = Interlocked.Increment(ref index);
+                if (ix == 4)
+                {
+                    throw new TestException();
+                }
+
+                await Task.Delay(1000, ct);
+                if (ct.IsCancellationRequested)
+                {
+                    return;
+                }
+                if (failMode != Fail.Default)
+                {
+                    Assert.True(false);
+                }
+            }, 8, failMode);
+
+            ParallelizedSummary result = null;
+
+            if (failMode == Fail.Default)
+            {
+                await Assert.ThrowsAsync<TestException>(async () =>
+                {
+                    result = await pTask;
+                });
+                Assert.Equal(10, index);
+            }
+            else
+            {
+                result = await pTask;
+                Assert.False(result.IsSucess);
+                Assert.False(result.IsCancelled);
+                Assert.NotEmpty(result.Exceptions);
+            }
+        }
+
+        private sealed class TestException : Exception {}
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Make this exception 'public'. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS3871)